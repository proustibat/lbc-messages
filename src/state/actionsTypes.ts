export enum messagesActionTypes {
  INIT = '@@INIT',
  REQUEST_MESSAGES = 'REQUEST_MESSAGES',
  REQUEST_MESSAGES_SUCCESS = 'REQUEST_MESSAGES_SUCCESS',
  REQUEST_MESSAGES_FAILURE = 'REQUEST_MESSAGES_FAILURE',
}
