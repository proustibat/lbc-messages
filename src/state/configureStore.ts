import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rootReducer from './reducers';
import { RootStateOrAny } from 'react-redux';

const middleware = [thunk];

const configureStore = (preloadedState?: RootStateOrAny) => {
  return createStore(
    rootReducer,
    preloadedState,
    composeWithDevTools(applyMiddleware(...middleware)),
  );
};

export default configureStore;
