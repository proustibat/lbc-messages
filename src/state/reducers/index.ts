import { combineReducers } from 'redux';
import messages, {MessagesReducerType} from './messagesReducer';

export type RootReducersType = {
  messages: MessagesReducerType
};
export default combineReducers({
  messages,
});
