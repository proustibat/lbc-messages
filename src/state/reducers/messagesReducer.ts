import { AxiosResponse } from 'axios';
import { messagesActionTypes as Actions } from '../actionsTypes';
import { MessageProps } from '../../components/Message';

export type MessagesReducerType = {
  items: MessageProps[];
  loading: boolean;
};

export const initialState: MessagesReducerType = {
  items: [],
  loading: false,
};

const messagesReducer = (
  state = initialState,
  action: {
    type: Actions;
    payload?: { data?: unknown; request?: AxiosResponse };
  },
) => {
  switch (action.type) {
    case Actions.REQUEST_MESSAGES:
      return {
        ...state,
        loading: true,
      };
    case Actions.REQUEST_MESSAGES_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case Actions.REQUEST_MESSAGES_SUCCESS:
      return {
        ...state,
        items: (action?.payload?.data as { messages: MessageProps[] })
          ?.messages,
        loading: false,
      };
    case Actions.SEND_MESSAGE_SUCCESS:
      return {
        ...state,
        items: [
          ...state.items,
          action.payload?.data
        ],
        loading: false,
      };
    default:
      return state;
  }
};

export default messagesReducer;
