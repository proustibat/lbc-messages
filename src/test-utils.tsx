import React, { ReactNode } from 'react';
import configureStore, { MockStore } from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { RootReducersType } from './state/reducers';

export const getMockedStore = (
  initialState: RootReducersType = { messages: { items: [], loading: false } },
) => {
  const createMockStore = configureStore([thunk]);
  const mockStore: MockStore = createMockStore(initialState);
  return mockStore;
};

export const withProvider = (WrappedComponent: React.ComponentType<any>) => (
  initialState: RootReducersType,
) => (props: any) => {
  return (
    <Provider store={getMockedStore(initialState)}>
      <WrappedComponent {...props} />
    </Provider>
  );
};

export const withRouter = (children: any) => {
  return (
    <MemoryRouter>
      <Route>{children}</Route>
    </MemoryRouter>
  );
};
