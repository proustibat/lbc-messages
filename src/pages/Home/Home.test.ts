import React from 'react';
import { render } from '@testing-library/react';
import { Home } from './index';
import { messages } from '../../fixtures/messages';
import { RootReducersType } from '../../state/reducers';
import { withProvider, withRouter } from '../../test-utils';

const stateLoadingMessages: RootReducersType = {
  messages: { items: [], loading: true },
};

const stateWithMessages: RootReducersType = {
  messages: { items: messages, loading: false },
};

const stateWithoutMessages: RootReducersType = {
  messages: { items: [], loading: false },
};

describe('Home view', () => {
  describe.each([
    ['the store has messages', stateWithMessages],
    ['the store has not any messages', stateWithoutMessages],
    ['the store is loading messages', stateLoadingMessages],
  ])('given %s', (_, state) => {
    it('should render the view correctly by default', () => {
      // Given
      const ProvidedHome = withProvider(Home)(state);

      // When
      const { container } = render(withRouter(ProvidedHome));

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
