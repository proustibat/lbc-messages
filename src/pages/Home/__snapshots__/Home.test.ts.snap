// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home view given the store has messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiBox-root MuiBox-root-1 messageList_marginTop-16"
    >
      <div
        class="MuiBox-root MuiBox-root-2 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
      >
        <h6
          class="MuiTypography-root counter_textTransform-uppercase MuiTypography-subtitle1 MuiTypography-alignRight"
          data-testid="counter"
        >
          messageWithCounter
        </h6>
        <div
          class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="message-accordion"
        >
          <div
            aria-controls="111-content"
            aria-disabled="false"
            aria-expanded="true"
            class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
            id="111-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiAccordionSummary-content Mui-expanded"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                />
              </svg>
              <p
                class="MuiTypography-root heading_fontSize-0.9375rem heading_marginLeft-8 MuiTypography-body1"
                data-testid="title"
              >
                title 1
              </p>
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-entered"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="111-header"
                  id="111-content"
                  role="region"
                >
                  <div
                    class="MuiAccordionDetails-root"
                  >
                    <p
                      class="MuiTypography-root body_fontSize-0.9375rem body_color-rgba(0, 0, 0, 0.54) MuiTypography-body1"
                    >
                      body 1
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="message-accordion"
        >
          <div
            aria-controls="222-content"
            aria-disabled="false"
            aria-expanded="true"
            class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
            id="222-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiAccordionSummary-content Mui-expanded"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                />
              </svg>
              <p
                class="MuiTypography-root heading_fontSize-0.9375rem heading_marginLeft-8 MuiTypography-body1"
                data-testid="title"
              >
                title 2
              </p>
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-entered"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="222-header"
                  id="222-content"
                  role="region"
                >
                  <div
                    class="MuiAccordionDetails-root"
                  >
                    <p
                      class="MuiTypography-root body_fontSize-0.9375rem body_color-rgba(0, 0, 0, 0.54) MuiTypography-body1"
                    >
                      body 2
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="message-accordion"
        >
          <div
            aria-controls="333-content"
            aria-disabled="false"
            aria-expanded="true"
            class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded"
            id="333-header"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiAccordionSummary-content Mui-expanded"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                />
              </svg>
              <p
                class="MuiTypography-root heading_fontSize-0.9375rem heading_marginLeft-8 MuiTypography-body1"
                data-testid="title"
              >
                title 3
              </p>
            </div>
            <div
              aria-disabled="false"
              aria-hidden="true"
              class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </div>
          <div
            class="MuiCollapse-container MuiCollapse-entered"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <div
                  aria-labelledby="333-header"
                  id="333-content"
                  role="region"
                >
                  <div
                    class="MuiAccordionDetails-root"
                  >
                    <p
                      class="MuiTypography-root body_fontSize-0.9375rem body_color-rgba(0, 0, 0, 0.54) MuiTypography-body1"
                    >
                      body 3
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-3 plusButton_position-fixed plusButton_left-0 plusButton_bottom-16 plusButton_width-100%"
    >
      <button
        aria-label="add"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Home view given the store has not any messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiBox-root MuiBox-root-4 messageList_marginTop-16"
    >
      <div
        class="MuiBox-root MuiBox-root-5 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
      >
        <h6
          class="MuiTypography-root counter_textTransform-uppercase MuiTypography-subtitle1 MuiTypography-alignCenter"
          data-testid="counter"
        >
          empty-list
        </h6>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-6 plusButton_position-fixed plusButton_left-0 plusButton_bottom-16 plusButton_width-100%"
    >
      <button
        aria-label="add"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Home view given the store is loading messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <div
      class="MuiBox-root MuiBox-root-7 messageList_marginTop-16"
    >
      <div
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
        role="progressbar"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
        />
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
        />
      </div>
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text skeleton_marginBottom-8 MuiSkeleton-wave"
      />
    </div>
    <div
      class="MuiBox-root MuiBox-root-8 plusButton_position-fixed plusButton_left-0 plusButton_bottom-16 plusButton_width-100%"
    >
      <button
        aria-label="add"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
